pkg/build/kube/tarbits.go

```diff
@@ -0,0 +1,425 @@
+// tarRoot is a local file path.
+//
+// The given tarRoot may also take the format tarRoot::localPath where the
+// tarRoot is separated from the local path where the remote files will be
+// downloaded.
+//
+// If no explicit local path is provided then remote files are downloaded to
+// "$HOME/.kind/tarbits/VERSION/GOOS/GOARCH" where VERSION is the value
+// extracted from the "version" file in the "kubernetes.tar.gz" tarball.
+//
+// If a file is already present on disk with the same size as the remote
+// equivalent, the file will not be downloaded again.
+func NewTarBits(tarRoot string) (bits Bits, err error) {
+	tarBits := &TarBits{}
+
+	if strings.HasPrefix(tarRoot, "file://") {
+		tarBits.tarRoot = strings.Replace(tarRoot, "file://", "", 1)
+		tarBits.tbitsType = tarBitsLocal
+	} else if strings.HasPrefix(tarRoot, "http:") ||
+		strings.HasPrefix(tarRoot, "https:") {
+		tarBits.remoteURI = tarRoot
+		tarBits.tbitsType = tarBitsHTTP
+	} else if ok, _ := regexp.MatchString(
+		`^(v\d+\.\d+\.\d+[\w.\-+]*)(::.+)?$`, tarRoot); ok {
+		tarBits.remoteURI = fmt.Sprintf("%s/release/%s", releaseURIPrefix, tarRoot)
+		tarBits.tbitsType = tarBitsSemVer
+	} else if strings.HasPrefix(tarRoot, "ci/") {
+		tarBits.remoteURI = fmt.Sprintf("%s/%s", ciURIPrefix, tarRoot)
+		tarBits.tbitsType = tarBitsCI
+	} else if strings.HasPrefix(tarRoot, "release/") {
+		tarBits.remoteURI = fmt.Sprintf("%s/%s", releaseURIPrefix, tarRoot)
+		tarBits.tbitsType = tarBitsRelease
+	} else {
+		tarBits.tarRoot = tarRoot
+		tarBits.tbitsType = tarBitsLocal
+	}
+
+	// Check to see if there's a local path encoded into the remote URI.
+	if tarBits.tarRoot == "" {
+		parts := strings.Split(tarBits.remoteURI, "::")
+		tarBits.remoteURI = parts[0]
+		if len(parts) > 1 {
+			tarBits.tarRoot = parts[1]
+		}
+	}
+
+	return tarBits, nil
+}
+
+// Build implements Bits.Build
+func (b *TarBits) Build() error {
+	// If this is a CI or Release URI then check to see if the URI is just a
+	// pointer to the actual version.
+	if b.tbitsType == tarBitsCI || b.tbitsType == tarBitsRelease {
+		if err := b.getRealURIForCIOrRelease(); err != nil {
+			return err
+		}
+	}
+
+	// If the files are not local then download them.
+	if b.tbitsType != tarBitsLocal {
+
+		// Allow the OS and architecture to be configurable based on the
+		// GOOS and GOARCH environment variables rather than the GOOS and
+		// GOARCH of the running process.
+		goos := os.Getenv("GOOS")
+		goarch := os.Getenv("GOARCH")
+		if goos == "" {
+			goos = util.GetOS()
+		}
+		if goarch == "" {
+			goarch = util.GetArch()
+		}
+
+		// Get the version from the remote "kubernetes.tar.gz"
+		version, err := b.getVersion()
+		if err != nil {
+			return err
+		}
+		fmt.Printf("Building version %s\n", version)
+
+		// Get the location where to the remote files are cached locally.
+		if b.tarRoot == "" {
+			currentUser, err := user.Current()
+			if err != nil {
+				return err
+			}
+			b.tarRoot = path.Join(
```

Comment:
probably best to not modify the original b.tarRoot field here and use a temp var instead.

Reply:
Hi @neolit123,

Hmm, I think I disagree. This case is only activated when `tarRoot == ""`.